Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp

```diff
@@ -0,0 +1,1320 @@
+      } catch (std::invalid_argument &) {
+      }
+      m_IntensityScalingIdx.push_back(parameterIndex("IntensityScaling" + si));
+    }
+  }
+};
+
+} // namespace
+
+/// Constructor
+CrystalFieldFunction::CrystalFieldFunction()
+    : IFunction(), m_nControlParams(0), m_nControlSourceParams(0),
+      m_dirtyTarget(true) {}
+
+// Evaluates the function
+void CrystalFieldFunction::function(const FunctionDomain &domain,
+                                    FunctionValues &values) const {
+  updateTargetFunction();
+  if (!m_target) {
+    throw std::logic_error(
+        "FunctionGenerator failed to generate target function.");
+  }
+  m_target->function(domain, values);
+}
+
+/// Set the source function
+/// @param source :: New source function.
+void CrystalFieldFunction::setSource(IFunction_sptr source) const {
+  m_source = source;
+}
+
+size_t CrystalFieldFunction::getNumberDomains() const {
+  if (!m_target) {
+    buildTargetFunction();
+  }
+  if (!m_target) {
+    throw std::runtime_error("Failed to build target function.");
+  }
+  return m_target->getNumberDomains();
+}
+
+std::vector<IFunction_sptr>
+CrystalFieldFunction::createEquivalentFunctions() const {
+  checkTargetFunction();
+  std::vector<IFunction_sptr> funs;
+  auto &composite = dynamic_cast<CompositeFunction &>(*m_target);
+  for (size_t i = 0; i < composite.nFunctions(); ++i) {
+    auto fun = composite.getFunction(i);
+    auto cfun = dynamic_cast<CompositeFunction *>(fun.get());
+    if (cfun) {
+      cfun->checkFunction();
+    }
+    funs.push_back(fun);
+  }
+  return funs;
+}
+
+/// Set i-th parameter
+void CrystalFieldFunction::setParameter(size_t i, const double &value,
+                                        bool explicitlySet) {
+  checkSourceFunction();
+  if (i < m_nControlParams) {
+    m_control.setParameter(i, value, explicitlySet);
+    m_dirtyTarget = true;
+  } else if (i < m_nControlSourceParams) {
+    m_source->setParameter(i - m_nControlParams, value, explicitlySet);
+    m_dirtyTarget = true;
+  } else {
+    checkTargetFunction();
+    m_target->setParameter(i - m_nControlSourceParams, value, explicitlySet);
+  }
+}
+
+/// Set i-th parameter description
+void CrystalFieldFunction::setParameterDescription(
+    size_t i, const std::string &description) {
+  checkSourceFunction();
+  if (i < m_nControlParams) {
+    m_control.setParameterDescription(i, description);
+  } else if (i < m_nControlSourceParams) {
+    m_source->setParameterDescription(i - m_nControlParams, description);
+  } else {
+    checkTargetFunction();
+    m_target->setParameterDescription(i - m_nControlSourceParams, description);
+  }
+}
+
+/// Get i-th parameter
+double CrystalFieldFunction::getParameter(size_t i) const {
+  checkSourceFunction();
+  checkTargetFunction();
+  if (i < m_nControlParams) {
+    return m_control.getParameter(i);
+  } else if (i < m_nControlSourceParams) {
+    return m_source->getParameter(i - m_nControlParams);
```

Comment:
should we also check if the index is valid?

Reply:
We do it already. `m_nControlSourceParams` is set to the total number of parameters in `m_control` and `m_source`. `m_source->getParameter` makes also its own check.

Reply:
:)